pageserver/src/page_cache.rs

```diff
@@ -96,183 +123,460 @@ impl Slot {
+            PAGE_CACHE.get().unwrap().remove_mapping(self_key);
+            self.inner.key = None;
         }
     }
+}
 
-    pub fn lookup(
+enum ReadBufResult<'a> {
+    Found(PageReadGuard<'a>),
+    NotFound(PageWriteGuard<'a>),
+}
+
+enum WriteBufResult<'a> {
+    Found(PageWriteGuard<'a>),
+    NotFound(PageWriteGuard<'a>),
+}
+
+impl PageCache {
+    //
+    // Section 1: Public interface functions for looking up and memorizing materialized page
+    // versions in the page cache
+    //
+
+    /// Look up a materialized page version.
+    ///
+    /// The 'lsn' is an upper bound, this will return the latest version of
+    /// the given block, but not newer than 'lsn'. Returns the actual LSN of the
+    /// returned page.
+    pub fn lookup_materialized_page(
         &self,
         tenant_id: ZTenantId,
         timeline_id: ZTimelineId,
         rel_tag: RelTag,
         blknum: u32,
         lsn: Lsn,
-    ) -> Option<(Lsn, LookupGuard)> {
-        let map_key = MapKey::new(tenant_id, timeline_id, rel_tag, blknum);
+    ) -> Option<(Lsn, PageReadGuard)> {
+        let mut map_key = MapKey::MaterializedPage {
+            hash_key: MaterializedPageHashKey {
+                tenant_id,
+                timeline_id,
+                rel_tag,
+                blknum,
+            },
+            lsn,
+        };
 
-        let version: Version = {
-            let map = self.map.read().unwrap();
-            let versions = map.get(&map_key)?;
+        if let Some(guard) = self.try_lock_for_read(&mut map_key) {
+            let MapKey::MaterializedPage { hash_key: _, lsn } = map_key;
+            Some((lsn, guard))
+        } else {
+            None
+        }
+    }
 
-            // TODO binary_search?
-            versions.iter().rev().find(|v| v.lsn <= lsn)?.clone()
+    ///
+    /// Store an image of the given page in the cache.
+    ///
+    pub fn memorize_materialized_page(
+        &self,
+        tenant_id: ZTenantId,
+        timeline_id: ZTimelineId,
+        rel_tag: RelTag,
+        blknum: u32,
+        lsn: Lsn,
+        img: &[u8],
+    ) {
+        let map_key = MapKey::MaterializedPage {
+            hash_key: MaterializedPageHashKey {
+                tenant_id,
+                timeline_id,
+                rel_tag,
+                blknum,
+            },
+            lsn,
         };
 
-        let slot = &self.slots[version.slot_idx];
-        let inner = slot.inner.read().unwrap();
-
-        if !inner.is_for_key(&map_key, lsn) {
-            return None;
+        match self.lock_for_write(&map_key) {
+            WriteBufResult::Found(write_guard) => {
+                // We already had it in cache. Another thread must've put it there
+                // concurrently. Check that it had the same contents that we
+                // replayed.
+                assert!(*write_guard == img);
+            }
+            WriteBufResult::NotFound(mut write_guard) => {
+                write_guard.copy_from_slice(img);
```

Comment:
Need to mark_valid. Currently the cache doesn't store anything, the benchmark has regressed to ~85 s